Painter: set up world in painter initialization

molly-moen opened the pull request:
In the painter we need a reference to the "World", which is the state of the neighborhood. World is a singleton with a reference to a Grid. On painter init, we get the World, and if it does not already have a grid, we load the grid from file. This will load from `serialized_maze.txt`. If it already has a grid, we don't change it.

Since pyodide reuses the same loaded packages between runs, we also need to ensure we reset the neighborhood between runs. I added a `reset_neighborhood` function that is run on teardown that does that.

Now you can print the grid for the neighborhood and see updates (if you make them manually 😄) You can also see that the grid is getting reset between runs in the screenshot (note that the corner is no longer blue and the paint bucket count goes back to 4)
![Screenshot 2025-01-16 at 3 31 26 PM](https://github.com/user-attachments/assets/df2e0089-a803-4aff-ad9c-419fdd974a2b)

I added `World` as an export from the neighborhood module to support resetting it. That technically makes it easier for students to mess with the world, but it would only impact their own program. They also could already do what I did in the screenshot and access the world via the painter, since all python fields are public.

## Links
- jira ticket: [CT-959](https://codedotorg.atlassian.net/browse/CT-959)


## Testing story
Tested locally and added unit tests.



## PR Checklist:

<!--
  The final step! Before you create your PR, double-check that everything is in order.
  Change [ ] to [X] during creation to check boxes.
-->

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

hannahbergam commented on python/pythonlab/neighborhood/tests/test_painter.py:
```diff
@@ -1,4 +1,11 @@
+def setUp():
+  world = World()
+  # Set up the world to be from a string rather than trying to load a file.
+  world.set_grid_from_string(SAMPLE_MAZE)
```
I imagine set_grid_from_string and set_grid_from_file come with World? Or we've defined them elsewhere?

fisher-alice replied:
Yes - https://github.com/code-dot-org/code-dot-org/blob/c3ae88126cf6a64b5e891bff8646c3ac74751268/python/pythonlab/neighborhood/neighborhood/support/world.py#L20-L24

Non-blocking: similar question to [this comment](https://github.com/code-dot-org/code-dot-org/pull/63356#discussion_r1919269297) about `GridFactory` static methods...wondering about adding these to a `grid_helpers` module.

molly-moen replied:
I think I'll clean these up in my next pr just to keep things moving

hannahbergam commented on python/pythonlab/pythonlab_setup/tests/test_reset_neighborhood.py:
```diff
@@ -0,0 +1,9 @@
+  world = World()
+  world.set_grid_from_string('[[{"tileType":1,"value":0,"assetId":0},{"tileType":1,"value":1,"assetId":0}],[{"tileType":0,"value":0,"assetId":0},{"tileType":1,"value":0,"assetId":0}]]')
+  assert world.grid is not None
+  reset_neighborhood()
```
Thanks for adding this test!

hannahbergam reviewed (approved):
LGTM! I asked one question but it isn't blocking

fisher-alice commented on python/pythonlab/neighborhood/neighborhood/__init__.py:
```diff
@@ -1 +1,3 @@
 from .painter import Painter as Painter
+# We export World so we can clear the world in pythonlab_setup
```
Thanks for the comment!

fisher-alice commented on python/pythonlab/pythonlab_setup/pythonlab_setup/reset_neighborhood.py:
```diff
@@ -0,0 +1,11 @@
+
+def reset_neighborhood():
+  """
+  Reset the neighborhood so the world will be re-generated on next painter initialization.
```
Just confirming my understanding. So we call on `reset_neighborhood` every time we call on `setup_pythonlab` which is called on when we load `pythonlab`, and when we click on 'Run'. Is that correct?
For 'on next painter initialization', does this refer to when we initialize a [`Neighborhood` instance](https://github.com/code-dot-org/code-dot-org/blob/2839e7ce38b9d15e447697cb5b97a8046447a385/apps/src/codebridge/MiniAppPreview/NeighborhoodPreview.tsx#L34) (or when we initialize a `Painter`)?

molly-moen replied:
This is the next time a user calls `Painter()` in code

fisher-alice reviewed (approved):
Looks good! Just left a couple small questions. Thanks!

molly-moen commented:
fyi @fisher-alice and @hannahbergam I moved the reset function to teardown rather than setup because I thought we always ran setup before running code but we actually run it on initializing python and then after teardown. It wasn't necessary to run this on initialization since the neighborhood will definitely not have a world yet, so I moved it.

hannahbergam commented:
> fyi @fisher-alice and @hannahbergam I moved the reset function to teardown rather than setup because I thought we always ran setup before running code but we actually run it on initializing python and then after teardown. It wasn't necessary to run this on initialization since the neighborhood will definitely not have a world yet, so I moved it.

Makes sense to me!
